baremetal/metal3machine_manager.go

```diff
@@ -831,6 +833,14 @@ func (m *MachineManager) chooseHost(ctx context.Context) (*bmov1alpha1.BareMetal
 		}
 		reqs = append(reqs, *r)
 	}
+	if m.Metal3Machine.Spec.FailureDomain != nil {
+		r, err := labels.NewRequirement(FailureDomainLabelPrefix, selection.Equals, []string{*m.Metal3Machine.Spec.FailureDomain})
+		if err != nil {
+			m.Log.Error(err, "Failed to create FailureDomain MatchLabel requirement, not choosing host")
+			return nil, nil, err
+		}
+		reqs = append(reqs, *r)
+	}
```

Comment:
@lentzi90 Sorry, I'm uninitated on conversion-gen. In the `make generate` output, I see

```
hack/tools/bin/conversion-gen \
                --input-dirs=./api/v1alpha5 \
                --output-file-base=zz_generated.conversion  --output-base=/Users/steven/oss/cluster-api-provider-metal3 \
                --go-header-file=./hack/boilerplate/boilerplate.generatego.txt
E0926 16:50:26.490736   38078 conversion.go:756] Warning: could not find nor generate a final Conversion function for github.com/metal3-io/cluster-api-provider-metal3/api/v1beta1.Metal3MachineSpec -> ./api/v1alpha5.Metal3MachineSpec
E0926 16:50:26.491121   38078 conversion.go:757]   the following fields need manual conversion:
E0926 16:50:26.491126   38078 conversion.go:759]       - FailureDomain
```

And in the generated conversion file

``` 
+++ b/api/v1alpha5/zz_generated.conversion.go
@@ -1246,6 +1246,7 @@ func autoConvert_v1beta1_Metal3ClusterStatus_To_v1alpha5_Metal3ClusterStatus(in
        out.LastUpdated = (*v1.Time)(unsafe.Pointer(in.LastUpdated))
        out.FailureReason = (*errors.ClusterStatusError)(unsafe.Pointer(in.FailureReason))
        out.FailureMessage = (*string)(unsafe.Pointer(in.FailureMessage))
+       // WARNING: in.FailureDomains requires manual conversion: does not exist in peer-type
        out.Ready = in.Ready
        // WARNING: in.Conditions requires manual conversion: does not exist in peer-type
        return nil
@@ -1715,14 +1716,10 @@ func autoConvert_v1beta1_Metal3MachineSpec_To_v1alpha5_Metal3MachineSpec(in *v1b
        out.MetaData = (*corev1.SecretReference)(unsafe.Pointer(in.MetaData))
        out.NetworkData = (*corev1.SecretReference)(unsafe.Pointer(in.NetworkData))
        out.AutomatedCleaningMode = (*string)(unsafe.Pointer(in.AutomatedCleaningMode))
+       // WARNING: in.FailureDomain requires manual conversion: does not exist in peer-type
        return nil
 }
```

But I am not sure where to make this manual conversion. I read a bit into it, and found in our v1alpha5 `doc.go`

```
// +k8s:conversion-gen=github.com/metal3-io/cluster-api-provider-metal3/api/v1beta1
```

Which looks right... but do I need to add annotations to the M3Cluster and M3Machine types in v1alpha5 as well?

If you could help point me in the right direction, I would appreciate it! :) Thanks!

Reply:
Sure!
Since you are adding a new field here to v1beta1 that does not exist in v1alpha5, conversion-gen does not know what to do with it. We cannot really do anything about it, other than just dropping the field and "taking the responsibility" for it :sweat_smile: . We do that by creating a conversion function and then just calling the automatically generated function, similar to this:

https://github.com/metal3-io/cluster-api-provider-metal3/blob/cde6c510d3f2811b44a8bc528c5420bba3767cca/api/v1alpha5/conversion.go#L219-L232

In other words, you need to add something like

```go
func Convert_v1beta1_Metal3MachineSpec_To_v1alpha5_Metal3MachineSpec(in *v1beta1.Metal3MachineSpec, out *Metal3MachineSpec, s apiconversion.Scope) error {
	// FailureDomain was added in v1beta1 so there is nothing we can do to convert it.
	// It will simply be dropped when converting to v1alpha5.
	return autoConvert_v1beta1_Metal3MachineSpec_To_v1alpha5_Metal3MachineSpec(in, out, s)
}
```

Reply:
Ah okay, thank you! I should have looked at this file, maybe I would have figured it out on my own. One more question though, why are we converting from v1beta1 to v1alpha5? Shouldn't these conversions be forward-looking?

Reply:
We actually convert both ways. :slightly_smiling_face:  This is a common pattern, and it works like this: There is one version that the controller handles, that is called the "hub" version. Then there can be any number of other supported versions called "spoken" versions. We add conversion functions to convert between the hub and the spoken versions (not between spoken versions directly, everything is always converted to/from the hub).

This system allows the controller to support any number of API versions simultaneously as long as we can convert them to/from the hub. The user can work with v1alpha5 as long as it is supported even though the controller is actually working with v1beta1. If the user asks for an object using v1alpha5, the webhook automatically converts it before returning it. Similarly, if the user creates an object using v1alpha5, it will be automatically converted and stored as v1beta1.

Kube-builder has some nice documentation around this if you want to read more: https://book.kubebuilder.io/multiversion-tutorial/conversion-concepts

Reply:
Thank you for explaining! And for the link. Cheers